# simpLAI
Local ancestry inference from small reference panels

simpLAI is a window-based approach that leverages information on the matching of haplotypes from admixed and reference individuals to infer the ancestry across the genome.
The program performs three types of inferences. In the first two modes, the ancestry assigned to a target haplotype of predefined size (-s) is that of the reference population that displays the lowest average number of mismatches with the target (mean) or that includes the chromosome displaying the lowest number of mismatches with the target (min). The third mode accounts for intra-population ancestry switches that may have occurred within a haplotype through recombination (rec). Each haplotype segment consists of n polymorphic sites (-n) and the number of mismatches between a target and each chromosome of the reference population is computed on subsets of t linked polymorphic sites (assumed to be a non-recombining segment). The path consisting of the combination of segments (potentially identified on different chromosomes) among the n polymorphic sites showing the smallest number of mismatches within each reference population is selected to compare reference populations. The reference with the shortest path is selected for the ancestry of the target haplotype. The increment for the sliding window is defined in base pairs (-i) for the mean and min mode and in number of polymorphic sites (-m) for the rec mode.
